### PR TITLE
gh-111569: Fix critical sections test on WebAssembly

### PR DESCRIPTION
**Pull request by colesbury**
This adds a macro `Py_CAN_START_THREADS` that corresponds to the Python function test.support.threading_helper.can_start_thread(). Some WebAssembly platforms do not have a working pthread implementation.

This macro is used to guard the critical sections C API tests that require a working threads implementation.

* Issue: gh-111569

### CODE REVIEW

**Comment by colesbury** [comment]
!buildbot wasm

**Comment by colesbury** [comment]
!buildbot wasm

File: Include/pyport.h
Comment (brettcannon) [inline Include/pyport.h]: Unfortunately it's trickier than that because you _can_ do a WASI build with experimental threading support (the "experimental" bit is on the WASI side, not our side); see the `--with-wasm-pthreads` for turning the support on.

https://github.com/python/cpython/blob/6f09f69b7f85962f66d10637c3325bbb2b2d9853/configure.ac#L2123-L2127

**Review by brettcannon — CHANGES_REQUESTED** [review]
Unfortunately there's more subtlety to thread support under WASI as it's possible via `--with-wasm-pthreads`, but it's opt-in (until threads fully land in the WebAssembly standard and support trickles down through the tooling).

File: Include/pyport.h
Comment (colesbury) [reply Include/pyport.h]: @brettcannon, do you think the general approach here is a reasonable? Previously, I had been just checking for thread support on the Python side, but that's a bit awkward for these C API tests.

I'm a bit confused by your comment that you can do a **WASI** build with threading support. I see that there is Emscripten support for pthreads. That's already handled by the `__EMSCRIPTEN_PTHREADS__` check on line 476.

The Python equivalent assumes that **WASI** never supports threads:
https://github.com/python/cpython/blob/6f09f69b7f85962f66d10637c3325bbb2b2d9853/Lib/test/support/threading_helper.py#L232-L233

I see that there is a WASI proposal for threads (https://github.com/WebAssembly/wasi-threads), but that looks like it's still marked "phase 1".

Comment (brettcannon) [reply Include/pyport.h]: > The Python equivalent assumes that **WASI** never supports threads

The threading support isn't well-tested because I haven't set up a buildbot yet and made it priority to more thoroughly test it. 😅

> I see that there is a WASI proposal for threads (https://github.com/WebAssembly/wasi-threads), but that looks like it's still marked "phase 1".

Yep, and it's staying at that phase because threading is being taken up by the WebAssembly CG at the W3C which acts as an upstream to WASI. That proposal hit I think phase 4 last month (it's definitely moving forward regardless of whether I got the number right), hence why this whole "threads in WASI" thing continues to be experimental both in WASI and for us.

I think we need to make a decision about free-threading and WASI. Do we want to try and make it work with the experimental threading support that's there, or do we say it's not worth it and wait on the more official support to land?

Comment (colesbury) [reply Include/pyport.h]: > I think we need to make a decision about free-threading and WASI.

My 2¢ is "no, not for now", but I also think that's only tangentially related to this issue. The test breakage is not happening in a free-threaded (`--disable-gil`) build. Even if we don't support the combination of `--disable-gil` and wasm, I'd still like a way to determine at compile time (in C code) if `PyThread_start_new_thread` will actually launch a thread (or is just a no-op stub). 

I think we might be able to just use `HAVE_PTHREAD_STUBS` here instead of defining a new `Py_CAN_START_THREADS`.

I'm hoping to run this on the wasm buildbots, but I think the runner is backed up... probably because the `test_critical_sections_threads` test hangs until the test suite times out.

Comment (brettcannon) [reply Include/pyport.h]: 👍 if the `HAVE_PTHREADS_STUBS` solution works.

Comment (brettcannon) [reply Include/pyport.h]: And I think the buildbots have calmed down: https://buildbot.python.org/all/#/builders?tags=wasm

Comment (colesbury) [reply Include/pyport.h]: Hmmm... `HAVE_PTHREAD_STUBS` did not seem to work.

Comment (colesbury) [reply Include/pyport.h]: It appears that when built without thread support WASI uses `HAVE_PTHREAD_STUBS`, but Emscripten stubs pthreads internally.

**Comment by colesbury** [comment]
Ok, I think this is ready for review now. I went back to the `Py_CAN_START_THREADS` approach, but check for `HAVE_PTHREAD_STUBS` and the Emscripten combiation. WASI support for threads is indicated by the absence/presence of `HAVE_PTHREAD_STUBS`. Emscripten support for threads is indicated by `__EMSCRIPTEN_PTHREADS__`.

**Review by brettcannon — APPROVED** [review]

**Comment by brettcannon** [comment]
!buildbot wasm

**Comment by brettcannon** [comment]
Thanks, @colesbury !

**Pull request merged**
